analyzers/src/SonarAnalyzer.CSharp/Rules/AsyncVoidMethod.cs

```diff
@@ -77,22 +75,22 @@ public sealed class AsyncVoidMethod : SonarDiagnosticAnalyzer
             || HasAnyMsTestV1AllowedAttribute(methodSymbol);
 
         private static bool IsUsedAsEventHandler(MethodDeclarationSyntax methodDeclaration) =>
-            methodDeclaration.FirstAncestorOrSelf<ClassDeclarationSyntax>() is ClassDeclarationSyntax parentClass
-            && parentClass.DescendantNodes()
+            methodDeclaration.FirstAncestorOrSelf((SyntaxNode x) => x.IsKind(SyntaxKind.ClassDeclaration) || x.IsKind(SyntaxKindEx.RecordDeclaration)) is { } parentDeclaration
```

Comment:
```suggestion
            methodDeclaration.FirstAncestorOrSelf<TypeDeclarationSyntax>(x => x.IsAnyKind(SyntaxKind.ClassDeclaration, SyntaxKindEx.RecordDeclaration)) is { } parentDeclaration
```
I think this can also do the job and I expect it to do a better job as only for TypeDeclarationSyntax will it do the checks.